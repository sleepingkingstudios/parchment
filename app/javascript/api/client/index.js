import {
  applyMiddleware,
  selectMiddleware,
} from 'api/middleware/utils';
import {
  valueOrDefault,
} from 'utils/object';

import generateActions from './actions';
import generateHooks from './hooks';
import generateInitialState from './initialState';
import generateReducer from './reducer';
import generateRequest from './request';
import generateSelector from './selector';

const wrapReducer = ({ middleware, reducer }) => applyMiddleware(
  reducer,
  selectMiddleware(middleware, 'handleAction'),
);

const buildClient = (options) => {
  const {
    namespace,
    url,
  } = options;
  const data = valueOrDefault(options.data, {});
  const method = valueOrDefault(options.method, 'GET');
  const middleware = valueOrDefault(options.middleware, []);

  const actions = generateActions({ namespace });
  const initialState = generateInitialState({ data });
  const performRequest = generateRequest({
    actions,
    method,
    middleware,
    namespace,
    url,
  });
  const hooks = generateHooks({ performRequest });
  const reducer = wrapReducer({
    middleware,
    reducer: generateReducer({ actions, initialState }),
  });
  const selector = generateSelector({ namespace });

  return {
    actions,
    hooks,
    method,
    middleware,
    namespace,
    performRequest,
    reducer,
    selector,
    url,
  };
};

export default buildClient;
