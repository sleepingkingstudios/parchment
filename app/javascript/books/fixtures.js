import { load } from '../fixtures';

/* eslint-disable-next-line import/prefer-default-export */
export const booksData = load('books');
