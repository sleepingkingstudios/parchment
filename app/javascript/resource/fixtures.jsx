import React from 'react';

export const Block = () => (<div />);
export const Form = () => (<div />);
export const Table = () => (<div />);

export const CreatePage = () => (<div />);
export const IndexPage = () => (<div />);
export const LandingPage = () => (<div />);
export const ShowPage = () => (<div />);
export const UpdatePage = () => (<div />);

export const fixtures = {
  create: {
    component: CreatePage,
    exact: true,
    options: { baseUrl: '/widgets', resourceName: 'widgets' },
    path: '/create',
    reducer: ((state = { widget: {} }) => state),
  },
  index: {
    component: IndexPage,
    exact: true,
    options: { baseUrl: '/widgets', resourceName: 'widgets' },
    path: '/',
    reducer: ((state = { widgets: [] }) => state),
  },
  show: {
    component: ShowPage,
    exact: true,
    options: { baseUrl: '/widgets', resourceName: 'widgets' },
    path: '/:id',
    reducer: ((state = { widget: {} }) => state),
  },
  update: {
    component: UpdatePage,
    exact: true,
    options: { baseUrl: '/widgets', resourceName: 'widgets' },
    path: '/:id/update',
    reducer: ((state = { widget: {} }) => state),
  },
};
