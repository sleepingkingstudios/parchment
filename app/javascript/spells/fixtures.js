/* eslint-disable-next-line import/prefer-default-export */
export const spellsData = [
  {
    id: 'e17bca01-24d3-48d3-a308-099dd6dca373',
    level: 0,
    name: 'Embers',
    school: 'Evocation',
  },
  {
    id: 'e3641970-fc98-4674-b873-5532d89d9f26',
    level: 3,
    name: "Dragon's Breath",
    school: 'Evocation',
  },
  {
    id: 'a74d8c21-607d-4a3f-8cfd-2bbb48ec196a',
    level: 9,
    name: 'Wrath of the Titans',
    school: 'Evocation',
  },
];
