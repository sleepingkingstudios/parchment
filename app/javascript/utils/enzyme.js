import { valueOrDefault } from './object';
import { safeCapitalize } from './string';

export const extractInjectedComponent = (component) => {
  const injected = component.shallow().find('ComponentWithInjectedProps');

  return injected.exists() ? injected : component;
};

export const toHaveFormControl = expect => (received, expectedControl, withProps) => {
  const formControl = received.find(expectedControl);
  const expectedProps = valueOrDefault(withProps, {});

  expect(formControl).toExist();
  expect(formControl).toHaveProp(expectedProps);

  return { pass: true, message: () => '' };
};

export const toHaveFormGroup = expect => (received, expectedGroup, fn) => {
  const formGroup = received.find(expectedGroup);

  expect(formGroup).toExist();

  fn(extractInjectedComponent(formGroup).shallow());

  return { pass: true, message: () => '' };
};

export const toHaveFormInput = expect => (received, props) => {
  const {
    form,
    inputType,
    prop,
  } = props;

  expect(received).toHaveProp({ prop });

  const formInput = received.find(`${safeCapitalize(prop)}Input`);

  expect(formInput).toExist();
  expect(formInput).toHaveProp({ form });

  const shallowInput = formInput.shallow();

  expect(shallowInput).toExist();
  expect(shallowInput).toHaveDisplayName(valueOrDefault(inputType, 'FormInput'));

  return { pass: true, message: () => '' };
};
