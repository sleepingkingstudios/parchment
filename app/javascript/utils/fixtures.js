/* eslint-disable-next-line import/prefer-default-export */
export const trimMultilineLiteral = (text) => {
  if (!text) { return ''; }

  const paragraphs = text.split('\n\n');

  return paragraphs
    .map((paragraph) => {
      const lines = paragraph.split('\n');

      return lines.map(line => line.trim()).join(' ');
    })
    .join('\n\n');
};
