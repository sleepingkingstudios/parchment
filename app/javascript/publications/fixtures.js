import { load } from '../fixtures';

const isOfficial = (publication) => {
  const { playtest, publisherName } = publication;

  return !playtest && publisherName === 'Wizards of the Coast';
};

const rawData = load('publications');

/* eslint-disable-next-line import/prefer-default-export */
export const publicationsData = rawData
  .map(publication => Object.assign(
    {},
    publication,
    {
      official: isOfficial(publication),
    },
  ));
