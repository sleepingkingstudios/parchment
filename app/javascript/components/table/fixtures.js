import { humanizeList } from '../../utils/array';

export const formatCrewCapacity = (item) => {
  const count = item.crewCapacity;

  if (count === 1) { return '1 Kerbal'; }

  return `${count} Kerbals`;
};

export const formatExperiments = (item) => {
  humanizeList(item.experiments, { default: 'None' });
};

export const columns = [
  {
    label: 'Name',
    prop: 'name',
  },
  {
    label: 'Crew Capacity',
    prop: 'crewCapacity',
    value: formatCrewCapacity,
  },
  {
    label: 'Experiments',
    prop: 'experiments',
  },
];

export const rocketsData = [
  {
    id: '83f5e9e8-92df-4203-bc2e-4381370c2186',
    name: 'Apprentice',
    crewCapacity: 1,
    experiments: [
      'Crew Report',
      'Mystery Goo',
      'Temperature Scan',
      'Pressure Scan',
    ],
  },
  {
    id: '4bd7454c-3b15-4047-8d57-dfa831dac58e',
    name: 'Sorcerer',
    crewCapacity: 2,
    experiments: [
      'Crew Report',
      'Materials Study',
      'Mystery Goo',
      'Temperature Scan',
      'Pressure Scan',
    ],
  },
  {
    id: '2fead083-d9ea-4f04-949e-d8c00e800cbb',
    name: 'Warlock',
    crewCapacity: 3,
    experiments: [
      'Crew Report',
      'Materials Study',
      'Mystery Goo',
      'Temperature Scan',
      'Pressure Scan',
      'Gravity Scan',
    ],
  },
];
