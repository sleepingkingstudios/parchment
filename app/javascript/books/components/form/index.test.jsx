import React from 'react';
import { shallow } from 'enzyme';

import BookForm from './index';

import { booksData } from '../../fixtures';
import { INITIALIZED } from '../../../api/status';

describe('<BookForm />', () => {
  const onChangeAction = jest.fn(
    ({ propName, value }) => ({ payload: { propName, value } }),
  );
  const onSubmitAction = jest.fn(() => ({ ok: true }));
  const data = { book: booksData[0] };
  const errors = {};
  const defaultProps = {
    data,
    onChangeAction,
    onSubmitAction,
    status: INITIALIZED,
    errors,
  };
  const form = {
    data,
    errors,
    path: ['book'],
    onChangeAction,
    onSubmitAction,
  };

  it('should render a form', () => {
    const rendered = shallow(<BookForm {...defaultProps} />);

    expect(rendered).toHaveDisplayName('Form');
    expect(rendered).toHaveProp({ form });
  });

  it('should render the title field', () => {
    const rendered = shallow(<BookForm {...defaultProps} />);
    const input = rendered.find('TitleField');

    expect(input).toExist();
    expect(input).toHaveProp({ form });
  });

  it('should render the slug field', () => {
    const rendered = shallow(<BookForm {...defaultProps} />);
    const input = rendered.find('SlugField');

    expect(input).toExist();
    expect(input).toHaveProp({ form });
  });

  it('should render the publisher name field', () => {
    const rendered = shallow(<BookForm {...defaultProps} />);
    const input = rendered.find('PublisherNameField');

    expect(input).toExist();
    expect(input).toHaveProp({ form });
  });

  it('should render the publication date field', () => {
    const rendered = shallow(<BookForm {...defaultProps} />);
    const input = rendered.find('PublicationDateField');

    expect(input).toExist();
    expect(input).toHaveProp({ form });
  });

  it('should render the cancel button', () => {
    const rendered = shallow(<BookForm {...defaultProps} />);
    const input = rendered.find('CancelButton');

    expect(input).toExist();
    expect(input).toHaveProp({ form });
  });

  it('should render the submit button', () => {
    const rendered = shallow(<BookForm {...defaultProps} />);
    const input = rendered.find('SubmitButton');

    expect(input).toExist();
    expect(input).toHaveProp({ form });
  });

  it('should match the snapshot', () => {
    const rendered = shallow(<BookForm {...defaultProps} />);

    expect(rendered).toMatchSnapshot();
  });
});
