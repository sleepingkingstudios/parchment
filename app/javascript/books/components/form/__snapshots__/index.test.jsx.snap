// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookForm /> should match the snapshot 1`] = `
<Form
  className="book-form"
  form={
    Object {
      "data": Object {
        "book": Object {
          "abbreviation": "sff",
          "id": "38a6666f-d3eb-4853-bd0a-add4d8ec369a",
          "playtest": true,
          "publicationDate": "2013-08-15",
          "publisherName": "Spelljammer Publishing",
          "slug": "spectres-flumph-fortress",
          "title": "Spectres of Flumph Fortress",
        },
      },
      "errors": Object {},
      "onChangeAction": [MockFunction],
      "onSubmitAction": [MockFunction],
      "path": Array [
        "book",
      ],
    }
  }
>
  <FormRow
    align="left"
    className=""
    verticalAlign="top"
  >
    <TitleField
      colWidth="8"
      form={
        Object {
          "data": Object {
            "book": Object {
              "abbreviation": "sff",
              "id": "38a6666f-d3eb-4853-bd0a-add4d8ec369a",
              "playtest": true,
              "publicationDate": "2013-08-15",
              "publisherName": "Spelljammer Publishing",
              "slug": "spectres-flumph-fortress",
              "title": "Spectres of Flumph Fortress",
            },
          },
          "errors": Object {},
          "onChangeAction": [MockFunction],
          "onSubmitAction": [MockFunction],
          "path": Array [
            "book",
          ],
        }
      }
    />
    <SlugField
      colWidth="4"
      form={
        Object {
          "data": Object {
            "book": Object {
              "abbreviation": "sff",
              "id": "38a6666f-d3eb-4853-bd0a-add4d8ec369a",
              "playtest": true,
              "publicationDate": "2013-08-15",
              "publisherName": "Spelljammer Publishing",
              "slug": "spectres-flumph-fortress",
              "title": "Spectres of Flumph Fortress",
            },
          },
          "errors": Object {},
          "onChangeAction": [MockFunction],
          "onSubmitAction": [MockFunction],
          "path": Array [
            "book",
          ],
        }
      }
    />
    <PublisherNameField
      colWidth="8"
      form={
        Object {
          "data": Object {
            "book": Object {
              "abbreviation": "sff",
              "id": "38a6666f-d3eb-4853-bd0a-add4d8ec369a",
              "playtest": true,
              "publicationDate": "2013-08-15",
              "publisherName": "Spelljammer Publishing",
              "slug": "spectres-flumph-fortress",
              "title": "Spectres of Flumph Fortress",
            },
          },
          "errors": Object {},
          "onChangeAction": [MockFunction],
          "onSubmitAction": [MockFunction],
          "path": Array [
            "book",
          ],
        }
      }
    />
    <PublicationDateField
      colWidth="4"
      form={
        Object {
          "data": Object {
            "book": Object {
              "abbreviation": "sff",
              "id": "38a6666f-d3eb-4853-bd0a-add4d8ec369a",
              "playtest": true,
              "publicationDate": "2013-08-15",
              "publisherName": "Spelljammer Publishing",
              "slug": "spectres-flumph-fortress",
              "title": "Spectres of Flumph Fortress",
            },
          },
          "errors": Object {},
          "onChangeAction": [MockFunction],
          "onSubmitAction": [MockFunction],
          "path": Array [
            "book",
          ],
        }
      }
    />
  </FormRow>
  <FormRow
    align="right"
    className=""
    verticalAlign="top"
  >
    <CancelButton
      colWidth="3"
      form={
        Object {
          "data": Object {
            "book": Object {
              "abbreviation": "sff",
              "id": "38a6666f-d3eb-4853-bd0a-add4d8ec369a",
              "playtest": true,
              "publicationDate": "2013-08-15",
              "publisherName": "Spelljammer Publishing",
              "slug": "spectres-flumph-fortress",
              "title": "Spectres of Flumph Fortress",
            },
          },
          "errors": Object {},
          "onChangeAction": [MockFunction],
          "onSubmitAction": [MockFunction],
          "path": Array [
            "book",
          ],
        }
      }
      isUpdate={false}
    />
    <SubmitButton
      colWidth="3"
      disabled={false}
      form={
        Object {
          "data": Object {
            "book": Object {
              "abbreviation": "sff",
              "id": "38a6666f-d3eb-4853-bd0a-add4d8ec369a",
              "playtest": true,
              "publicationDate": "2013-08-15",
              "publisherName": "Spelljammer Publishing",
              "slug": "spectres-flumph-fortress",
              "title": "Spectres of Flumph Fortress",
            },
          },
          "errors": Object {},
          "onChangeAction": [MockFunction],
          "onSubmitAction": [MockFunction],
          "path": Array [
            "book",
          ],
        }
      }
      isUpdate={false}
      status="initialized"
    />
  </FormRow>
</Form>
`;
